splitio/push/manager.py

```diff
@@ -247,3 +283,226 @@ def _handle_connection_end(self):
+        :type blocking: bool
+        """
+        if not self._running:
+            _LOGGER.warning('Push manager does not have an open SSE connection. Ignoring')
+            return
+
+        self._running = False
+        await self._processor.update_workers_status(False)
+        self._status_tracker.notify_sse_shutdown_expected()
+        self._next_refresh.cancel()
+        await self._sse_client.stop(blocking)
+
+    async def _event_handler(self, event):
+        """
+        Process an incoming event.
+
+        :param event: Incoming event
+        :type event: splitio.push.sse.SSEEvent
+        """
+        try:
+            parsed = self._get_parsed_event(event)
+            handle = self._event_handlers[parsed.event_type]
+        except (KeyError, EventParsingException):
+            _LOGGER.error('Parsing exception or no handler for message of type %s', parsed.event_type)
+            _LOGGER.debug(str(event), exc_info=True)
+            return
+
+        try:
+            await handle(parsed)
+        except Exception:  # pylint:disable=broad-except
+            _LOGGER.error('something went wrong when processing message of type %s',
+                          parsed.event_type)
+            _LOGGER.debug(str(parsed), exc_info=True)
+
+    async def _token_refresh(self):
+        """Refresh auth token."""
+        _LOGGER.info("retriggering authentication flow.")
+        self.stop(True)
+        await self._trigger_connection_flow()
+
+    async def _trigger_connection_flow(self):
+        """Authenticate and start a connection."""
+        try:
+            token = await self._auth_api.authenticate()
+        except APIException:
+            _LOGGER.error('error performing sse auth request.')
+            _LOGGER.debug('stack trace: ', exc_info=True)
+            await self._feedback_loop.put(Status.PUSH_RETRYABLE_ERROR)
+            return
+
+        if not token.push_enabled:
+            await self._feedback_loop.put(Status.PUSH_NONRETRYABLE_ERROR)
+            return
+        self._telemetry_runtime_producer.record_token_refreshes()
+
+        _LOGGER.debug("auth token fetched. connecting to streaming.")
+        self._status_tracker.reset()
+        self._running = True
+        if await self._sse_client.start(token):
+            _LOGGER.debug("connected to streaming, scheduling next refresh")
+            await self._setup_next_token_refresh(token)
+            self._running = True
+            self._telemetry_runtime_producer.record_streaming_event((StreamingEventTypes.CONNECTION_ESTABLISHED, 0,  get_current_epoch_time_ms()))
+
+    async def _setup_next_token_refresh(self, token):
+        """
+        Schedule next token refresh.
+
+        :param token: Last fetched token.
+        :type token: splitio.models.token.Token
+        """
+        if self._next_refresh is not None:
+            self._next_refresh.cancel()
+        self._next_refresh = TimerAsync((token.exp - token.iat) - _TOKEN_REFRESH_GRACE_PERIOD,
```

Comment:
The async class use different timer class, as the threading one Timer is based on threading. The TimerAsync is based on asyncio lib.

Reply:
yes, i mean the `(token.exp - token.iat) - _TOKEN_REFRESH_GRACE_PERIOD`. we can calculate that in a function and pass it to each timer class instead of having hthe calcuation twice, in case we ever change it's formula

Reply:
will do